### PR TITLE
validator: validating age + tests

### PR DESCRIPTION
**Pull request by martinmilev**
Ready for review.

### CODE REVIEW

File: src/com/clouway/agevalidator/UserHandler.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+public class UserHandler {
+  private List<User> users = new LinkedList<>();
+
+  public void newUser(User user, Validator v) {
+    if(ageValidate(user,v) && !users.contains(user)){
+      users.add(user);
+    }
+  }
+
+  public boolean ageValidate(User user, Validator v) {
```
Comment (mgenov) [inline src/com/clouway/agevalidator/UserHandler.java]: Should not be public.

File: src/com/clouway/agevalidator/UserHandler.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.List;
+
+/**
+ * @Author Martin Milev
+ * @Date 08.06.16.
+ */
+public class UserHandler {
+  private List<User> users = new LinkedList<>();
+
+  public void newUser(User user, Validator v) {
```
Comment (mgenov) [inline src/com/clouway/agevalidator/UserHandler.java]: It is better Validator to be passed to the constructor, cause If I want to use UserHandler I have to pass him Validator. 

The UserHandler need to ensure that only valid users will pass. 

With your implementation I'm able to do the following: 

```java
UserHandler handler = new UserHandler(); 
handler.newUser(new User("john", 20), new Validator() {
   // it's better to be validateAge()
   public boolean ageValidate(User user) {
      return true;
   }
  
   public boolean userIsAdult(User user) {
      return true;
   }
}
```

And with this code I will be able to register any user that I want.

File: src/com/clouway/agevalidator/AgeValidator.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package com.clouway.agevalidator;
+
+/**
+ * @Author Martin Milev
+ * @Date 08.06.16.
+ */
+public class AgeValidator implements Validator {
```
Comment (mgenov) [inline src/com/clouway/agevalidator/AgeValidator.java]: Забележка: Да се тестват методите в изолация. Да не се създават реални имплементации на валидатора и класа за връзка с базата данни. Задачата да се реализира с TDD подход.

**Comment by mgenov** [comment]
This is what tasks says: 

3) Да се реализира service клас, който изълнява следните функционалности:

получава годините на човек под формата на String. Валидира този стринг посредством валидатор , като проверява дали стойността е валидна. Ако е валидна (валидни стойности: от 10 до 100), стойността може да бъде запазена в база данни.

извлича годините на човек от база данни. Ако стойността е по-голяма от 18- връща true. Ако е по-малка- връща false.

Забележка: Да се тестват методите в изолация. Да не се създават реални имплементации на валидатора и класа за връзка с базата данни. Задачата да се реализира с TDD подход.

I'm not able to find db interface for loading of users by there age.

File: src/com/clouway/agevalidator/UserHandler.java (hunk 1 above)
Comment (martinmilev) [inline src/com/clouway/agevalidator/UserHandler.java]: > извлича годините на човек от база данни. Ако стойността е по-голяма от 18- връща true. Ако е по-малка- връща false.

So I should have a public method like that one somewhere?

Comment (mgenov) [inline src/com/clouway/agevalidator/UserHandler.java]: Validator interface already has it.

File: src/com/clouway/agevalidator/AgeValidator.java (hunk 3 above)
Comment (martinmilev) [inline src/com/clouway/agevalidator/AgeValidator.java]: So I have to delete the implementations?

Comment (mgenov) [inline src/com/clouway/agevalidator/AgeValidator.java]: Yeah, they are not required for this task.

Comment (martinmilev) [inline src/com/clouway/agevalidator/AgeValidator.java]: Should I fake the db and the validator, or just leave it like that?

Comment (mgenov) [inline src/com/clouway/agevalidator/AgeValidator.java]: I don't see the service class here. You are doing the same mistake as before. Mocks are used for simulation of collaborator's behaviour not for testing the framework. 

As It's not clear I've added a sample class that shows the intention: https://github.com/clouway/teamspace/wiki/Testing-with-mocks

**Comment by martinmilev** [comment]
I've removed the implementations of the interfaces. Should I fake test them, or just leave it like that?

File: src/com/clouway/agevalidator/AgeValidator.java (hunk 3 above)
Comment (martinmilev) [inline src/com/clouway/agevalidator/AgeValidator.java]: Fixed it. Now the tests actually test something. Makes more sense that way.

File: src/com/clouway/agevalidator/UserRepository.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    this.userDB = db;
+    this.validator = v;
+  }
+  public void registerUser(User u) {
+    if(validator.ageValidate(u) || !userDB.userExists(u.name())) {
+      userDB.addUser(u);
+    }
+  }
+
+  public boolean userExist(User u){
```
Comment (mgenov) [inline src/com/clouway/agevalidator/UserRepository.java]: This method is not required.

File: src/com/clouway/agevalidator/UserRepository.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    return userDB.userExists(u.name());
+  }
+
+  public boolean isAdult(User u) {
```
Comment (mgenov) [inline src/com/clouway/agevalidator/UserRepository.java]: The signature of this method in the task description  "public boolean isAdult(String name)" 

So you have to ask the UserDB to load user by it's name and after that to check whether it's adult or not. 

Please look carefully and the task description. 
https://github.com/clouway/teamspace/wiki/Testing%20with%20mocks

File: src/com/clouway/agevalidator/UserRepository.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  private final UserDB userDB;
+  private final Validator validator;
+
+  public UserRepository(UserDB db, Validator v) {
+
+    this.userDB = db;
+    this.validator = v;
+  }
+
+  public boolean userExist(User u){
```
Comment (mgenov) [inline src/com/clouway/agevalidator/UserRepository.java]: userExists was the method that wasn't required !!!

Comment (mgenov) [inline src/com/clouway/agevalidator/UserRepository.java]: This is the signature in the task description for clearance. 
```java
// The service class
public class UserRepository {
  public UserRepository(UserDB userDB, Validator validator) {
    ..
  }
  public void registerUser(User user) {
    ...
  }

  public boolean isAdult(String name) {
       ````
  }
}
```

Comment (mgenov) [inline src/com/clouway/agevalidator/UserRepository.java]: I don't know how to specify it more clearly.

Comment (martinmilev) [inline src/com/clouway/agevalidator/UserRepository.java]: I think I got it right.

File: test/com/clouway/agevalidator/UserDBTest.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+import org.junit.Test;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * @Author Martin Milev
+ * @Date 09.06.16.
+ */
+public class UserDBTest {
```
Comment (mgenov) [inline test/com/clouway/agevalidator/UserDBTest.java]: Test for the UserRepository should be UserRepositoryTest and not UserDBTest.

**Comment by mgenov** [comment]
LGTM !

**Pull request merged**